### PR TITLE
Fix Selu Activation Function

### PR DESCRIPTION
**Pull request by chrisrothUT**
@attila-i-szabo noticed that jax.nn.selu loses it's self normalizing property with complex numbers. Namely, it applies `selu(x) `when it should be applying `selu(Re(x)) + 1j*selu(Im(x)) `. Making this small change improves performance tremendously. I introduce the correct non-linearity as `piecewise_selu`

### CODE REVIEW

**Comment by attila-i-szabo** [comment]
I wouldn't be surprised if noone in the machine learning community generalised SELU to complex numbers, there's not much on complex-valued neural networks anywhere. You could in principle fix this in a number of ways, treating the real and imag. parts as real numbers and applying SELU to them is just the most straightforward, all you have to ensure is that the `lecun_normal` initialiser is aware of complex numbers, to adjust the variance used

**Comment by gcarleo** [comment]
I think this makes a lot of sense! Just to understand : what are we currently doing for complex ReLu? Are we splitting real and imaginary part there or doing something different?

**Comment by chrisrothUT** [comment]
It looks like `jax.nn.relu` uses `jax.numpy.maximum` which might only compare the real part. I'll reimplement it as well

**Comment by chrisrothUT** [comment]
Maybe we should have a wrapper function that takes a non-linearity and implements it piecewise across the real and imaginary parts?

**Comment by attila-i-szabo** [comment]
That sounds like a good idea, it would remove a lot of duplicate code.

BTW, there is a holomorphic construction for ReLU: you take z in some part of the complex plane ([these people](https://openreview.net/forum?id=H1T2hmZAb) suggest using Re(z)+Im(z)>0 as the cutoff) and 0 in the rest. This is ofc different from applying ReLU to real and imag parts separately

**Comment by gcarleo** [comment]
Yes indeed the issue is that there are many different ways of generalizing these activations to the complex domain.... Component-wise real/imaginary seems a reasonable option but I personally don't like too much that the activation does not mix imaginary and real part. However I have no idea if mixing is a good idea after all... I think this should be studied more in detail in a paper and I would suggest that you guys name these activations accordingly so that there is no confusion (say, CSeLu or something!)...

**Comment by attila-i-szabo** [comment]
> there are many different ways of generalizing these activations to the complex domain

Of course, but I don't think there is any generally tractable scheme other than splitting real/imaginary parts, and the only more creative ones I've seen were basically complexifying ReLU...

> you guys name these activations accordingly so that there is no confusion

So going with Chris' earlier suggestion, maybe the best would be having a decorator going like
```python
def complexify(f):
    def cf(x):
        if iscomplexobj(x):
            return jax.lax.complex(f(x.real), f(x.imag))
        else:
            return f(x)
    return cf
```
so our new SELU would be `complexify(jax.nn.selu)`

> I think this should be studied more in detail in a paper

Hmm... the paper I linked compares a bunch of ReLU extensions, I'm not sure how much general theory one can invent beyond that

**Comment by gcarleo** [comment]
Yes I agree using the decorator is the right thing to do. My suggestion would be just to put these complexified activations in a separate namespace or have them start with a C (say...) otherwise there might be some ambiguity, for example the ReLu as it is now is defined correctly using Max on the abs of complex numbers but what you are proposing is a different version... Also you might want to define for example a complexified logcosh etc acting independtly on real and  and imag part etc etc

**Comment by gcarleo** [comment]
Unless I misunderstood what you were saying, maybe you were proposing to just add the decorator and let the users use it as they please? (that would also work well!)

**Comment by chrisrothUT** [comment]
Added the decorator `complexify` as well as the `C_selu` and `C_relu` nonlinearities. Also switched the `GCNN` default to `C_selu`.

**Comment by PhilipVinc** [comment]
I have a few remarks:

0) I believe that in NetKet we should keep only things and techniques that are somewhat established and/or published. Other _things_ might still end up under the netket umbrella organisation, (something like a netket_extras) but not inside netket itself. Part of the reason is that NetKet should be stable, and experimental things, by definition, change often.

1) I'm all in favour of improving things, but we had several PRs changing GCNN recently. We released 3.0 and we should keep it stable if possible. If we change the activation function, we are not being stable. With that said:
  - If current selu does completely make no sense for complex inputs, we can see this PR as a bug fix and then it's technically ok to change it
  - However, again, is this the good choice of activation function? We have no idea. You say nobody/almost nobody uses complex-valued selu. So is this well-understood/acknoledged enough to end up in netket itself?

2) I'd still like a more through bibliography search on this: there are few, but there *are* articles studying complex valued netwroks. That's how we argued for `lecun_***` initialisers to be implemented into jax core; there are a few articles studying and defying them. And selu are built around `lecun` so maybe someone actually studied them.

  - Moreover, if this is established, instead of putting those initialisers in netket, we can put them in jax itself: I would like that al our `complex number bugfixes` be upstreamed into jax/flax and keep in jax/flax only stuff that really makes only we are using and they will refuse. 

-- 
So in short:

 - I'm in favour of `netket.nn.complexify` (though maybe we could have a better name?) possibly using HashablePartial in order to keep the hash constant and avoid recompilation if possible
 - We can define *some* activation functions with it, but let's bike shed the name: I'm against `C_relu`. I'd be more in favour of `reim_relu` for example, or something else.
 - If you want to merge this PR in netket itself please do a small analysis showing that this activation is better than current and the alternative that is in the literature?

**Comment by femtobit** [comment]
> If current selu does completely make no sense for complex inputs, we can see this PR as a bug fix and then it's technically ok to change it

This is something that to my understanding can be argued, because, according to @chrisrothUT:

> @attila-i-szabo noticed that jax.nn.selu loses it's self normalizing property with complex numbers.

...which is the main property that sets `selu` apart from more common activations, as far as I understand. So in that sense I'd argue this can be seen as a bugfix. (Saying we use selu for the complex GCNN appears to me to be a bit misleading otherwise.)

**Comment by PhilipVinc** [comment]
>This is something that to my understanding can be argued, because, according to @chrisrothUT:

Yes, which is why I'm open to this. Though I'd like to insist in at least trying to upstream this to jax first

**Comment by PhilipVinc** [comment]
>...which is the main property that sets selu apart from more common activations. So in that sense I'd argue this can be seen as a bugfix. (Saying we use selu for the complex GCNN appears to me to be a bit misleading otherwise.)

Though don't we want the output variance to be 1? The current implementation gives 2, if I am not mistaken...

**Comment by femtobit** [comment]
> Though don't we want the output variance to be 1? The current implementation gives 2, if I am not mistaken...

I think this depends on whether the activation is used in a real or complex setting. If so, I'd say its better to be exactly equal to normal selu in the real case and accept the constant factor in the complex case (which should only be some kind of gauge freedom anyways, right?).

**Comment by attila-i-szabo** [comment]
> I'm in favour of `netket.nn.complexify` (though maybe we could have a better name?) 

`reim` or `real_imag`?

> possibly using HashablePartial in order to keep the hash constant and avoid recompilation if possible

good point, although we can mostly apply it to predefined functions like `*_relu` and defaults, we can at best tell the user to construct their own complexifications thusly

> If current selu does completely make no sense for complex inputs, we can see this PR as a bug fix and then it's technically ok to change it

So the whole point of SELU is that if you give it approximately unit Gaussian distributed inputs, it deforms the output to be closer to such a distribution, which makes deep networks more stable than Kaiming initialised RELU. Analytically continuing SELU (as it happens now) doesn't have this property. Nevertheless, it kinda works for simple enough problems, but then we could use any activation at all

>  You say nobody/almost nobody uses complex-valued selu.

Complex numbers are not a common itch to be scratched in machine learning, so this is not surprising. There are a few papers, but they usually come in the wake of very well-established things, like I've only ever come across ReLU generalisations.

> If you want to merge this PR in netket itself please do a small analysis showing that this activation is better than current and the alternative that is in the literature?

The reason @chrisrothUT and I came across this was that GCNNs kept blowing up for various frustrated magnets we work on; this problem went away when I started using Re-Im SELU. What "alternatives in the literature" do you mean?

> Though don't we want the output variance to be 1? The current implementation gives 2, if I am not mistaken...

That is true and it's the origin of some of the complications that I outline above about using `lecun`, so it would be better to change it.
I guess we could scale the input and the output by sqrt(2). If we want to save the runtime of that, we can try to work out how the SELU formula changes, but that might require us to interact with the black magic that goes on in the original SELU paper...

**Comment by PhilipVinc** [comment]
>good point, although we can mostly apply it to predefined functions like *_relu and defaults, we can at best tell the user to construct their own complexifications thusly

Yep, indeed I agree. but if we are exposing it to users, they might define their own complexities functions and this is where `HashablePartial` becomes useful. It's a tiny change and it makes life easier so why not?

> So the whole point of SELU is that if you give it approximately unit Gaussian distributed inputs, it deforms the output to be closer to such a distribution, which makes deep networks more stable than Kaiming initialised RELU.

yep! I'm convinced about this! 

> The reason @chrisrothUT and I came across this was that GCNNs kept blowing up for various frustrated magnets we work on; this problem went away when I started using Re-Im SELU. What "alternatives in the literature" do you mean?

For relu (not selu) I know that you can either apply it to real-imaginary separately, you can apply it to one quadrant of the complex space, you can apply it only to the magnitude... 

if the objective of complex-selu is to give you a more unit gaussian distributed input, is it okay to apply it separately to real and imaginary? don't you get a bias if the input data is correlated among the two directions?

> I guess we could scale the input and the output by sqrt(2). If we want to save the runtime of that, we can try to work out how the SELU formula changes, but that might require us to interact with the black magic that goes on in the original SELU paper...

`sqrt(2)` will be precompiled, and be a constant, so it's only a multiplication... so I don't think this is much worse.

**Comment by attila-i-szabo** [comment]
> if the objective of complex-selu is to give you a more unit gaussian distributed input, is it okay to apply it separately to real and imaginary? don't you get a bias if the input data is correlated among the two directions?

I don't know whether anybody studied this, but I think there are analogous situations in real-valued networks (e.g. a convolutional network will also have nontrivial correlations due to reused kernel entries). I don't know whether covariances like this are dissipated by SELU, but it might be discussed in the original paper, might have a look later.

> `sqrt(2)` will be precompiled, and be a constant, so it's only a multiplication... so I don't think this is much worse.

In that case it might be a good idea to add it to `complexify` like
```python
def complexify(f):
    def cf(x):
        if iscomplexobj(x):
            return jax.lax.complex(f(x.real * 2**0.5), f(x.imag * 2**0.5)) / 2**0.5
        else:
            return f(x)
    return cf
```
so we don't have to worry about normalisation differences between real and complex inputs (which would make properly initialising a GCNN painful)

**Comment by attila-i-szabo** [comment]
> why it would be important to do so.

If you have some large covariances between outputs of one layer, those might get reinforced by the next layer, leading to some large variances in the next layer output. With many enough layers this too leads to instabilities, so you wouldn't gain much compared to critically initialised RELU. For this very reason I think it must help dissipate covariances as well as deviations from the unit Gaußian of each marginal distribution. However, I don't know if anyone ever formally studied this

**Comment by chrisrothUT** [comment]
> If you have some large covariances between outputs of one layer, those might get reinforced by the next layer, leading to some large variances in the next layer output. With many enough layers this too leads to instabilities, so you wouldn't gain much compared to critically initialised RELU. For this very reason I think it must help dissipate covariances as well as deviations from the unit Gaußian of each marginal distribution. However, I don't know if anyone ever formally studied this

Actually this kind of thing may help with the ill-conditioned Fisher matrices we encounter later in training. Might be worth looking into if this kind of thing can be designed with a small amount of effort

I think no one has studied this because you actually want some covariance on typical neural network problems. For example on classification tasks you want to sparsify the data into categories.

**Comment by chrisrothUT** [comment]
I can't find `HashablePartial` anywhere in NetKet. Where is it located?

**Comment by PhilipVinc** [comment]
`nk.jax.HashablePartial`

**Comment by gcarleo** [comment]
concerning our favorite topic: naming things, after all I also tend to agree that 
> C_relu 

is not so nice... (C makes me think of the old compiled C++ modules of netket...) `ri_relu` might work (I'd keep a short name if possible)

**Comment by gcarleo** [comment]
This looks good to me, the only thing maybe is changing names to `complexify` as also suggested by @PhilipVinc (maybe at this point `ri_complexify` is a better name, but I am not sure

**Comment by PhilipVinc** [comment]
possible names for `complexify`

 - `complexify` (how? I vote against it.)
   - Though this is basically calling `jax.lax.complex(.., ..)` so maybe there's some reason to it.
 - `complexify_reim`
   - `complexify_realimag` seems to long
 - `reim_complexify`
 - `ri_complexify`

For activation funcitons, I'd use the same prefix/postfix.

- `reim_selu` 
- `ri_selu`


I like `reim_****` best

**Comment by femtobit** [comment]
For what it's worth, I agree with @PhilipVinc: `reim_` looks relatively clear to me (more so than `ri_`) and its good to have consistent prefixes. I'm happy with either choice, though.

**Comment by chrisrothUT** [comment]
I agree with everybody, I like the reim prefix. I'm not sure we even need the "complexify" name, lets just call it reim.

**Comment by PhilipVinc** [comment]
I actually quite like `complexify`, I just wanted to be explicit on how it happens. `reim_complexify` or `reim_elementwise` are more explicit than `reim` that is a bit more misterious, if alone. What do you think ?

**Comment by gcarleo** [comment]
It's all good I like reim too. I am not sure about specifying elewentwise, since that might feel like a vector function rather than the element you have in mind here, but it's not a big deal let's merge this!

**Comment by PhilipVinc** [comment]
ok, then names are settled.
Let me just quickly review. I think a docstring or two are missing and I don't understand the usage of `HashablePartial`

File: netket/nn/activation.py [hunk 1]
```diff
@@ -48,6 +48,21 @@
 from jax.numpy import cosh
 from jax.numpy import sinh
 
+from netket.jax import HashablePartial
+
+
+@HashablePartial
+def reim(f):
+    # Modifies a non-linearity to act seperately on the real and imaginary parts
```
Comment (PhilipVinc) [inline netket/nn/activation.py:56]: Make it a docstring.
Also add it to https://github.com/netket/netket/blob/master/docs/docs/api.rst so that it shows up in the documentation.

(BTW, also add reim_relu, reim_selu to the website)

File: netket/nn/activation.py [hunk 2]
```diff
@@ -48,6 +48,21 @@
...
+    def cf(x):
+        sqrt2 = jnp.sqrt(jnp.array(2, dtype=x.real.dtype))
+        if jnp.iscomplexobj(x):
+            return jax.lax.complex(f(sqrt2 * x.real), f(sqrt2 * x.imag)) / sqrt2
+        else:
+            return f(x)
+
+    return cf
```
Comment (PhilipVinc) [inline netket/nn/activation.py:64]: that's how to use it.

```suggestion
    def reim_activation(f, x):
        sqrt2 = jnp.sqrt(jnp.array(2, dtype=x.real.dtype))
        if jnp.iscomplexobj(x):
            return jax.lax.complex(f(sqrt2 * x.real), f(sqrt2 * x.imag)) / sqrt2
        else:
            return f(x)

    return HashablePartial(reim_activation, f)
```

File: netket/nn/activation.py [hunk 3]
```diff
@@ -63,19 +78,23 @@ def log_tanh(x):
     return jax.numpy.log(jax.numpy.tanh(x))
 
 
+reim_selu = reim(selu)
+reim_relu = reim(relu)
```
Comment (PhilipVinc) [inline netket/nn/activation.py:82]: add two short docstrings to the two functions

**Review by PhilipVinc — CHANGES_REQUESTED** [review]
Almost good! The only missing things are

- a few missing docstrings and links to the website.
- Fix the usage of hashablePartial
-  please add a line to CHANGELOG.md under new features mentioning `nk.nn.activation.reim` and the two new activations.
- a new line under breaking changes mentioning the activation of the GCNN changed.

both with a link to this pr

File: netket/nn/activation.py (hunk 3 above)
Comment (chrisrothUT) [reply netket/nn/activation.py]: I'm not sure how to do this, as all `reim` is a function that returns a function

Comment (PhilipVinc) [reply netket/nn/activation.py]: just below
```python
reim_selu=reim(selu)
"""
my beautiful docstring
"""
```

**Review by PhilipVinc — APPROVED** [review]

**Comment by PhilipVinc** [comment]
looks good to me.
can i merge @chrisrothUT ?

**Comment by chrisrothUT** [comment]
> can i merge @chrisrothUT ?

Go for it!
